imap_data_access/file_validation.py

```diff
@@ -444,11 +542,122 @@ def construct_path(self) -> Path:
         return spice_dir / subdir / self.filename
 
+    @staticmethod
+    def _matches_on_group(
+        regex_list: Iterable[re.Pattern], string_to_parse: str
+    ) -> re.Match:
+        """Determine the first regular expression that matches the provided string.
+
+        Parameters
+        ----------
+        regex_list: list
+            A list of compiled regular expressions to be applied in order
+        string_to_parse: str
+                The string to check against the provided regular expressions
+
+        Returns
+        -------
+            The first match that satisfies a regular expression found in regex_list
+        """
+        for regex in regex_list:
+            m = regex.match(string_to_parse)
+            if m is not None:
+                return m
+        return None
+
+    @staticmethod
+    def _extract_parts(
+        filename: Path | str,
+        parts: list[str],
+        transforms: dict | None = None,
+        handle_missing_parts: bool = False,
+    ) -> dict | None:
+        """Extract the parts of a file.
+
+        Parameters
+        ----------
+        filename: Path | str
+            A filename to extract parts from
+        parts: list[str]
+            A list of parts to be extracted from the filename
+        transforms: dict[str: func]
+            A dictionary of group->function (that takes 1 string) to
+            transform the string into some other type
+        handle_missing_parts: bool
+            If True, missing parts won't raise an exception and the result
+            set will contain None for the missing part.
+            If False, an IndexError is raised
+
+        Returns
+        -------
+            A dictionary of the parts that were requested or
+            None if there were no matches for the provided regex_list
+        """
+        if isinstance(filename, Path):
+            filename = filename.name
+
+        m = SPICEFilePath._matches_on_group(
+            SPICEFilePath.valid_spice_regexes, filename.lower()
+        )
+        if not m:
+            return None
+        ret_val = {}
+        for part in parts:
+            try:
+                if transforms is not None and part in transforms:
+                    ret_val[part] = transforms[part](m.group(part))
+                else:
+                    ret_val[part] = m.group(part)
+            except IndexError:
+                if handle_missing_parts:
+                    if transforms is not None and part in transforms:
+                        ret_val[part] = transforms[part](None)
+                    else:
+                        ret_val[part] = None
+                else:
+                    ret_val[part] = None
+
+        return ret_val
+
+    @staticmethod
+    def extract_filename_components(
+        filename: Path | str, parts: list[str]
+    ) -> dict | None:
+        """Extract all components from filename.
+
+        Will return a dictionary with the following keys:
```

Comment:
All that we are getting is type and version? What about all of the start_year and other parts that were extracted via the regex. (I think we'd prefer to not pass in parts explicitly and let it fall out automatically based on what was matched?)

I'm a bit lost in the regex handling here, so some examples might be helpful of an input and expected return in one or two specific cases.

In the other functions in this file we regex match and extract out the components directly. Does that not work with the multiple regex pattern?

Reply:
> Alternatively, I could just have the extract_filename_components return anything and everything it can based on the file name, and get rid of the specific "parts" parameter

Yes, I think this would be ideal. Then add in your additional derived type/version as necessary to those found parts.